### ISSUE TITLE
Flaky test: Stack not rebuilding updated packages in a snapshot

### ISSUE THREAD

**Issue by rvl** [cardano-foundation/cardano-wallet#2812]
# Context
 
Stack builds the snapshot once and caches it in `$STACK_ROOT/snapshots/SYSTEM/HASH/COMPILER_VER`.

The hash seems to be made from the snapshot's `name` attribute only, not from its contents.

Therefore, if we update a snapshot YAML, but don't update its name, Stack may try to use outdated snapshot builds, leading to obscure compile errors.

This happens both locally and on Buildkite CI.

# Test Case

Stack build.

# Failure / Counter-example

```
byron-spec-ledger          > configure
byron-spec-ledger          > Configuring byron-spec-ledger-0.1.0.0...
byron-spec-ledger          > build
byron-spec-ledger          > Preprocessing library for byron-spec-ledger-0.1.0.0..
byron-spec-ledger          > Building library for byron-spec-ledger-0.1.0.0..
byron-spec-ledger          > [ 1 of 16] Compiling Byron.Spec.Ledger.Core
byron-spec-ledger          > Failed to load interface for  Cardano.Crypto.Hash.Keccak256 
byron-spec-ledger          > There are files missing in the  cardano-crypto-class-2.0.0  package,
byron-spec-ledger          > try running 'ghc-pkg check'.
byron-spec-ledger          > Use -v (or `:set -v` in ghci) to see a list of the files searched for.
 
--  While building package byron-spec-ledger-0.1.0.0 (scroll up to its section to see the error) using:
      /build/cardano-wallet.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.2.1.0_ghc-8.10.5 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.2.1.0 build --ghc-options " -fdiagnostics-color=always"
    Process exited with code: ExitFailure 1
```

# Workarounds

1. Remove all stack caches.
2. Any time that something needs to be changed in the snapshot, change its name.

**Comment by Anviking** [comment]
> Therefore, if we update a snapshot YAML, but don't update its name, Stack may try to use outdated snapshot builds, leading to obscure compile errors.

> This happens both locally and on Buildkite CI.

If @jonathanknowles ran into this locally from current `master`, it seems to suggest this happens irregardless of whether the snapshot name is updated along side changes, as the name for the new snapshot in the recently merged #2801 _was_ a new and unique one.

**Comment by Anviking** [comment]
Should be fixed with #2831
